### DISCUSSION TITLE
dynamically switch UART IO Pins

### DISCUSSION THREAD

**Discussion by giancarlo64** [micropython/micropython#11581] [ESP32]
Hi,
I have a serial device connected trough UART with my ESP32. Only during communication I can keep the RS232 connected. In the rest of the time the connection must be on high impedance (Tristate) (like no connection on the Tx Pin).
So I need to switch dynamically the UART Tx Pin from OUT to High impedance and back. 
I have tried the following code :

#enable UART
uart1 = UART(1, baudrate=115200, tx=3, rx=4, bits=8, stop=1)
uart.write('rdstat\r')
sleep(1.5)
data = uart.read() 
size=len(data)
data = data[:size - 2]
rx_bytearray=ubinascii.unhexlify(data)

other code 
......
.......

#disable TX PIN 
p3=Pin(3,Pin.IN)
sleep(1.5)
#enable UART
uart1 = UART(1, baudrate=115200, tx=3, rx=4, bits=8, stop=1)
sleep(2) # wait to settle
uart.write('rdstat\r')
sleep(1.5)
data = uart.read() 
size=len(data)
data = data[:size - 2]
rx_bytearray=ubinascii.unhexlify(data)

after I get the following error
ValueError: odd-length string   or  ValueError: non-hex digit found

If I don't switch off pin 3  I get no errors
How can I put the TX of the UART in Tristate mode when not used?

Regards

Gian Carlo

**Comment by kjm1102** [comment]
You could try clearing the buffer with a uart.read() before the uart.write()

**Reply by giancarlo64** [reply]
Hi kjm1102,
only clearing the buffer doesn't solve the problem. 
After doing some test I found tat when I re-enable the UART with

uart1 = UART(1, baudrate=115200, tx=3, rx=4, bits=8, stop=1)
 after I have to wait at least 2 seconds to allow the UART to settle up


uart1 = UART(1, baudrate=115200, tx=3, rx=4, bits=8, stop=1)
#UART settle up
sleep(2)
#clear the buffer
uart.read()

This work fine
if I reduce the time  (sleep(1.5)) I start get the same problem

But I don't like this solution. Do someone know a better solution?

Gian Carlo

**Reply by kjm1102** [reply]
maybe try uart.init() in lieu of sleep(2) ?

**Reply by zcattacz** [reply]
Instead of disabling directly, can you try assigning a different unused pin to UART first, recover pin state before assign UART back to it ?
